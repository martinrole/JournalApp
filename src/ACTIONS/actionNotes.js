import { addDoc, collection, deleteDoc, doc, updateDoc } from "@firebase/firestore"
import Swal from "sweetalert2"
import { baseDatos } from "../FIREBASE/firebaseConfig"
import { fileUploadHelper } from "../HELPERS/fileUpload"
import { loadNotes } from "../HELPERS/loadNotes"
import { types } from "../TYPES/types"


export const newNoteAction = () => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth

        const newNote = {
            titulo: '',
            texto: '',
            fecha: new Date().getTime()
        }

        const docRef = await addDoc( collection(baseDatos, `${uid}/journal/notes`), newNote )
        //console.log('docRef: ', docRef);
        
        dispatch( activeNote(docRef.id, newNote) )
        dispatch( addNewNote(docRef.id, newNote) )
    }
}

export const loadNotesAction = ( uid ) => {
    return async ( dispatch ) => {
        const notes = await loadNotes( uid )
        dispatch( setNotes( notes ) )
    }
}

export const saveNoteAction = ( note ) => async ( dispatch, getState ) => {
        
    const { uid } = getState().auth

    //En los 3 puntos queda el resto de propiedades de la nota, y así se puede extraer el id para no repetirlo
    const {id, imgUrl = null, ...notaFinal } = note

    //Asi se envia a Firebase en la versiòn beta 9
    const noteRef = doc(baseDatos, `${uid}/journal/notes/${note.id}`)
    await updateDoc( noteRef, { imgUrl, ...notaFinal })

    dispatch( refreshNote( note.id, note ))
    Swal.fire('Saved', note.titulo, 'success' )
}

export const deleteNoteAction = ( id ) => async ( dispatch, getState ) => {

    const { uid } = getState().auth

    const  noteRef = doc(baseDatos, `${uid}/journal/notes/${id}`)
    await deleteDoc(noteRef)

    dispatch( deleteNote( id ) )
}

export const fileUpload = ( archivo ) => async ( dispatch, getState ) => {
    
    const activeNote = getState().notes.activeNote

    Swal.fire({
        title: 'Uploading...',
        text: 'Please wait...',
        allowOutsideClick: false,
        showConfirmButton: false,
        willOpen: () => {
            Swal.showLoading()
        }
    })

    
    const fileUrl = await fileUploadHelper( archivo )
    activeNote.imgUrl = fileUrl
    
    dispatch( saveNoteAction( activeNote ) )
    Swal.close()

}

export const addNewNote = ( id, nota ) => ({
    type: types.notesAdd,
    payload: { id, ...nota }
})

export const refreshNote = ( id, nota ) => ({
    type: types.notesUpdate,
    payload: { id, nota: { id, ...nota } }
})


export const activeNote = (id, nota) => ({
    type: types.notesActive,
    payload: { id, ...nota }
})

export const setNotes = ( nota ) => ({
    type: types.notesLoad,
    payload: nota
})

export const deleteNote = ( id ) => ({
    type: types.notesDelete,
    payload: id
})

export const clearNotes = () => ({
    type: types.notesClear
})
