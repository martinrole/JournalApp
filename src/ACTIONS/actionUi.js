import { types } from "../TYPES/types"

export const actionSetError = ( error ) => ({
    type: types.uiSetError,
    payload: error
})

export const actionRemoveError = () => ({
    type: types.uiRemoveError
})


export const actionLoading = (type) => ({
    type: type
})

