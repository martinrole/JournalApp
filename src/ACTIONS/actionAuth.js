import { autorizacion, googleAuth } from "../FIREBASE/firebaseConfig";
import { signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword, signOut } from 'firebase/auth';
import { types } from "../TYPES/types"
import { actionLoading, actionSetError } from "./actionUi";
import Swal from 'sweetalert2'
import { clearNotes } from "./actionNotes";

//Middleware:
export const actionLoginAsync = ( correo, contrasena ) => {
    //Esta funciÃ³n va a retornar un Callback:
    //el dispatch lo va  a entregar el thunk instalado
    return (dispatch) => {

        dispatch( actionLoading(types.uiLoadingOn) )

        return signInWithEmailAndPassword( autorizacion, correo, contrasena )
            .then( ({user}) => {
                dispatch( actionLogin( user.uid, user.displayName ) )
                dispatch( actionLoading(types.uiLoadingOff) )
            })
            .catch ( error => {
                dispatch( actionSetError(error.code) )
                Swal.fire('Error', error.code, 'error')
                dispatch( actionLoading(types.uiLoadingOff) )
            }) 
            
        }
}



export const actionRegisterAsync = ( nombre, correo, contrasena ) => {
    return ( dispatch ) => {

        return createUserWithEmailAndPassword( autorizacion, correo, contrasena )
            .then( async ( { user } ) => {
                await updateProfile( user, { displayName: nombre } )
                
                dispatch(
                    actionLogin( user.uid, user.displayName)
                )
            })
            .catch ( error => {
                dispatch( actionSetError(error.code) )
                Swal.fire('Error', error.code, 'error')
            })  
    }
}

export const actionGoogleLogin = () => {
    return ( dispatch ) => {
        signInWithPopup( autorizacion, googleAuth )
            .then( ( { user } ) => {
                dispatch(
                    actionLogin( user.uid, user.displayName)
                )

            })     
    }
}

export const actionLogOutAsync = () => {
    return async ( dispatch ) => {
        await signOut(autorizacion)

        dispatch( actionLogOut() )
        dispatch( clearNotes() )
    }
}

//Una forma de hacerlo:
// export const actionAuth = (uid, displayName) => {
//     return {
//         type: types.login,
//         payload: {
//             uid,
//             displayName
//         }
//     }
// }


//Segunda forma de hacerlo:
export const actionLogin = (uid, displayName) => ({  
    type: types.login,
    payload: {
        uid,
        displayName
    }
})

export const actionLogOut = () => ({
    type: types.logout
})