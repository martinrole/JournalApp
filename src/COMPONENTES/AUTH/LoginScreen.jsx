import React from 'react'
import { Link } from 'react-router-dom'
import { actionLoginAsync, actionGoogleLogin } from '../../ACTIONS/actionAuth'
import { useForm } from '../../HOOKS/useForm'
import { useDispatch, useSelector  } from 'react-redux'

const LoginScreen = () => {

    const dispatch = useDispatch()
    const { cargando } = useSelector(state => state.ui )

    const [ formValues, handleInputChange ] = useForm({
        correo: 'pruebaMartin@hotmail.com',
        contrasena: '12nj3nh4'
    })

    const { correo, contrasena } = formValues

    const handleSubmit = (e) => {
        e.preventDefault()
        dispatch( actionLoginAsync( correo,contrasena ) )
    }

    const handleGoogleLogin = () => {
        dispatch( actionGoogleLogin() )
    }    

    return (
        <>
            <h3 className="auth__title">LoginScreen</h3>
            
            <form onSubmit={ handleSubmit }>
                <input 
                    type="text" 
                    placeholder="Email" 
                    name="correo" 
                    className="auth__input" 
                    autoComplete="off"
                    value = { correo }
                    onChange={ handleInputChange }
                />
                <input 
                    type="password" 
                    placeholder="Password" 
                    name="contrasena" 
                    className="auth__input"
                    value= { contrasena }
                    onChange= { handleInputChange }
                />

                <button type="submit" className="btn btn-primario btn-bloque mt-1" disabled={ cargando }>{ cargando ? "Cargando" : "Login"}</button>
            </form>
           
            <div className="auth__social-networks">   
                <div className="google-btn" onClick={ handleGoogleLogin }>
                    <div className="google-icon-wrapper">
                        <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                    </div>
                    <p className="btn-text">
                        <b>Sign in with google</b>
                    </p>
                </div>
            </div>

            <Link to="/auth/register" className="auth__new-count"> Create new count</Link>

        </>
    )
}

export default LoginScreen
