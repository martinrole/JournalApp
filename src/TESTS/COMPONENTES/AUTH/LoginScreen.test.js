//Video de explicación de este archivo 291
import LoginScreen from "../../../COMPONENTES/AUTH/LoginScreen"
import { mount } from 'enzyme';
import { Provider } from "react-redux";

import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store' //ES6 modules
import thunk from 'redux-thunk';
import '@testing-library/jest-dom'
import { actionGoogleLogin, actionLoginAsync } from "../../../ACTIONS/actionAuth";

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {
    auth: {},
    ui: {
        cargando: false
    }
}

let store = mockStore( initState )
store.dispatch = jest.fn()

//Simula la ejecución de la función
jest.mock('../../../ACTIONS/actionAuth', () => ({
    actionGoogleLogin: jest.fn(),
    actionLoginAsync: jest.fn()
}))

const wrapper = mount( 
    <Provider store={ store }>
        <MemoryRouter>
            <LoginScreen/>
        </MemoryRouter>
    </Provider>
)

describe('Pruebas en componente LoginScreen', () => {

    beforeEach(() => {
        //Limpiar el store despuès de cada test
        store = mockStore( initState )
        jest.clearAllMocks()
    })

    test('Debe mostrar el componente correctamente', () => {
        expect( wrapper ).toMatchSnapshot()
    })

    test('Debe activar la acción de actionGoogleLogin', () => {

        wrapper.find('.google-btn').simulate('click')
        expect( actionGoogleLogin ).toHaveBeenCalledTimes(1)
    })

    test('Debe activar la acción de actionLoginAsync', () => {

        wrapper.find('form').simulate('submit')
        expect( actionLoginAsync ).toHaveBeenCalledTimes(1)
        expect( actionLoginAsync ).toHaveBeenCalledWith( 'pruebaMartin@hotmail.com','12nj3nh4')
    })

})