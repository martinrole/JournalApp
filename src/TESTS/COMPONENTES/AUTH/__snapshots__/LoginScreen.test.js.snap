// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente LoginScreen Debe mostrar el componente correctamente 1`] = `
Array [
  <h3
    className="auth__title"
  >
    LoginScreen
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="correo"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="pruebaMartin@hotmail.com"
    />
    <input
      className="auth__input"
      name="contrasena"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="12nj3nh4"
    />
    <button
      className="btn btn-primario btn-bloque mt-1"
      disabled={false}
      type="submit"
    >
      Login
    </button>
  </form>,
  <div
    className="auth__social-networks"
  >
    <div
      className="google-btn"
      onClick={[Function]}
    >
      <div
        className="google-icon-wrapper"
      >
        <img
          alt="google button"
          className="google-icon"
          src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
        />
      </div>
      <p
        className="btn-text"
      >
        <b>
          Sign in with google
        </b>
      </p>
    </div>
  </div>,
  <a
    className="auth__new-count"
    href="/auth/register"
    onClick={[Function]}
  >
     Create new count
  </a>,
]
`;
